pkg/ccl/backupccl/generative_split_and_scatter_processor_test.go

```diff
@@ -111,13 +113,129 @@ func TestRunGenerativeSplitAndScatterContextCancel(t *testing.T) {
+
+	const numAccounts = 1000
+	const localFoo = "nodelocal://0/foo"
+	ctx := context.Background()
+	tc, sqlDB, _, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts,
+		InitManualReplication)
+	defer cleanupFn()
+
+	st := cluster.MakeTestingClusterSettings()
+	evalCtx := eval.MakeTestingEvalContext(st)
+	evalCtx.NodeID = base.NewSQLIDContainerForNode(tc.Server(0).RPCContext().NodeID)
+
+	testDiskMonitor := execinfra.NewTestDiskMonitor(ctx, st)
+	defer testDiskMonitor.Stop(ctx)
+
+	// Set up the test so that the test context is canceled after the first entry
+	// has been processed by the generative split and scatterer.
+	s0 := tc.Server(0)
+	registry := tc.Server(0).JobRegistry().(*jobs.Registry)
+	execCfg := s0.ExecutorConfig().(sql.ExecutorConfig)
+	flowCtx := execinfra.FlowCtx{
+		Cfg: &execinfra.ServerConfig{
+			Settings:       st,
+			DB:             s0.InternalDB().(descs.DB),
+			JobRegistry:    registry,
+			ExecutorConfig: &execCfg,
+		},
+		EvalCtx:     &evalCtx,
+		Mon:         evalCtx.TestingMon,
+		DiskMonitor: testDiskMonitor,
+		NodeID:      evalCtx.NodeID,
+	}
+
+	sqlDB.Exec(t, `SET CLUSTER SETTING bulkio.backup.file_size = '1'`)
+	sqlDB.Exec(t, `BACKUP INTO $1`, localFoo)
+
+	backups := sqlDB.QueryStr(t, `SHOW BACKUPS IN $1`, localFoo)
+	require.Equal(t, 1, len(backups))
+	uri := localFoo + "/" + backups[0][0]
+
+	codec := keys.MakeSQLCodec(s0.RPCContext().TenantID)
+	backupTableDesc := desctestutils.TestingGetPublicTableDescriptor(tc.Servers[0].DB(), codec, "data", "bank")
+	backupStartKey := backupTableDesc.PrimaryIndexSpan(codec).Key
+
+	spec := makeTestingGenerativeSplitAndScatterSpec(
+		[]string{uri},
+		[]roachpb.Span{{
+			Key:    backupStartKey,
+			EndKey: backupStartKey.PrefixEnd(),
+		}},
+	)
+
+	// These split and scatterers will always fail the scatter and return 0 as the
+	// chunk destination.
+	chunkSplitScatterers := []splitAndScatterer{
+		&scatterAlwaysFailsSplitScatterer{},
+		&scatterAlwaysFailsSplitScatterer{},
+		&scatterAlwaysFailsSplitScatterer{},
+	}
+
+	// Fake some entries in the routing datum cache for nodes that we've scattered
+	// to already.
+	cache := routingDatumCache{
+		cache:   map[roachpb.NodeID]rowenc.EncDatum{8: {}, 9: {}},
+		nodeIDs: []roachpb.NodeID{8, 9},
+	}
+
+	// Large enough so doneScatterCh never blocks.
+	doneScatterCh := make(chan entryNode, 1000)
+	err := runGenerativeSplitAndScatter(ctx, &flowCtx, &spec, chunkSplitScatterers, chunkSplitScatterers, doneScatterCh, &cache)
+	require.NoError(t, err)
+
+	close(doneScatterCh)
+	var doneEntries []entryNode
+
+	for e := range doneScatterCh {
+		doneEntries = append(doneEntries, e)
+	}
+
+	// There are at least 10 splits from the original backed up bank table. Plus
+	// the entries from the system table, etc. Sanity check this.
+	require.GreaterOrEqual(t, len(doneEntries), 10)
+
+	// The failed scatter chunks should be scattered to the nodes that have been
+	// scattered to before and cached.
```

Comment:
is it worth also checking that the distribution is "roughly even"? I assume that's why you used a hash instead of rand.Int().

Reply:
I added some additional checking that there are some entries for each node, but the hashing doesn't make guarantees about the distribution either, but it does make the new assertion in the test consistent.

Reply:
Cool. If it starts to flake for some other reason, we can remove.